README.md

```diff
@@ -0,0 +1,170 @@
+# hwcart
+
+This library implements a hardware-aware Cartesian MPI communicator. 
+A standard MPI Cartesian communicator created with `MPI_Cart_create`
+allows the user to specify rank grid dimensions (`[nx, ny, nz]`, where
+`np=nx*ny*nz` is the total number of ranks) and periodicity. How the ranks
+are then distributed among the compute nodes, and how they are bound to
+compute cores depends on the MPI runtime.
+
+`hwcart` gives the user much finer control over how the ranks are
+placed on various levels of the memory hierarchy. Consider 2-socket compute nodes
+with AMD EPYC 7742 64-Core CPUs. 
+
+```
+$ hwloc-info
+depth 0:            1 Machine (type #0)
+ depth 1:           2 Package (type #1)
+  depth 2:          8 Group0 (type #12)
+   depth 3:         32 L3Cache (type #6)
+    depth 4:        128 L2Cache (type #5)
+     depth 5:       128 L1dCache (type #4)
+      depth 6:      128 L1iCache (type #9)
+       depth 7:     128 Core (type #2)
+        depth 8:    256 PU (type #3)
+Special depth -3:   8 NUMANode (type #13)
+
+```
+All cores have their private L1 and L2 caches, but there are 32 shared L3 caches. 
+This means that 4 cores share an L3 cache, 16 cores (4 L3 cache groups) share 
+the same NUMA node, and 4 NUMA-nodes are located on a single socket.
+
+`hwcart` uses [`hwloc`](https://www.open-mpi.org/projects/hwloc/) to understand
+the memory locality of the hardware and creates an equivalent of the MPI Cartesian
+communicator, in which the neigoboring ranks are placed close to each other in
+the complex memory hierarchy. This is done level by level, starring from the lowest
+desired level. Let's assume we want to start 128 ranks per compute node, on a total of
```

Comment:
I would make a new paragraph: example
and say that we are now looking at a 3D example